### PR TITLE
druid nested data column type

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
#12695

_TBD_

##### Key changed/added classes in this PR
 * `MyFoo`
 * `OurBar`
 * `TheirBaz`

<hr>

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by cheddar — APPROVED** [review]
I've reviewed this in private, seen it active and work in the wild, so am throwing my +1 on it.  Tests must pass, the description should be fleshed out and all of that before merge.

**Comment by gianm** [comment]
Nice work! Very cool. I just noticed this when you posted it on Slack. The first thing that comes to mind is questions about the intended maturity of this feature after initial contribution:

- Is it intentionally undocumented in this PR? Do you plan to add documentation?
- Are there any impediments to maintaining forwards compatibility, such that new versions of Druid will always be able to read JSON columns written by older versions? Do you foresee any reason we might want to break compatibility?
- Would you recommend we present this feature in its current state as experimental or production-ready, & why?

**Comment by clintropolis** [comment]
>Is it intentionally undocumented in this PR? Do you plan to add documentation?

I was planning to add documentation in a follow-up PR since I thought this one was already big enough 😅 

>Are there any impediments to maintaining forwards compatibility of the storage format, such that new versions of Druid will always be able to read JSON columns written by older versions? Do you foresee any reason we might want to break compatibility?

I modeled the column after existing Druid columns so most things are decorated with a version byte which should allow us to make changes in the future while still being able to continue reading the existing data. For the specific list of what is versioned: 
* `NestedDataColumnSerializer` for the complex column itself (currently on v3 actually, i removed the reader code for older versions from prototyping to get rid of dead code)
* `GlobalDictionaryEncodedFieldColumnWriter` which writes the nested columns and is currently re-using `DictionaryEncodedColumnPartSerde.VERSION` (i should probably decouple this at some point in the future...)
* `FixedIndexed` (building block used to store local to global dictionary mapping and long and double value dictionaries)
* `CompressedVariableSizedBlobColumnSerializer` (used to compress raw data)
* `CompressedBlockSerializer` (used internally by `CompressedVariableSizedBlobColumnSerializer`)

In the "Future work" section of #12695 I mention storage format as an area that we can iterate on in the future, the biggest things I have in mind right now are storing arrays of literal values as array typed columns instead of broken out as they currently are, as well as customization such as allowing skipping building indexes on certain columns or storing them all-together also probably falls into this. Nothing about the current code should block this afaik, nor should those future enhancements interfere with our ability to read data that is stored with the current versions of stuff, so long as we practice good version hygiene whenever we make changes.

> Would you recommend we present this feature in its current state as experimental or production-ready, & why?
This is a hard one to answer, though I am hesitant to call it production ready right from the start, I think the answer might vary a bit per use case.

The surface area here is huge since it essentially provides all of the normal Druid column functionality within these `COMPLEX<json>` columns, and I definitely won't claim this to be bug free. That said, quite a lot of internal testing has been done at this point, even at scale and with complicated nested schemas, which has allowed this codebase to be iterated on to get it to the place it currently is. There are some rough spots which I'm looking to improve in the near future, such as ingest time memory footprint, better array handling, etc, but I think if we get the documentation in a good enough state and can list out the limitations it could be used today.

The use cases I would feel most comfortable with are replacements for what can currently be done via flattening, meaning not heavily centered on nested arrays. I do have ideas of how to better support nested arrays and my goal is to allow arrays extracted from nested columns to be exposed as druid `ARRAY` types, but I am not there yet, so I'm not sure I would recommend most array use cases unless they are more like vectors which have expected lengths and array positions are known/meaningful (and such that most queries would be extracting specific array positions, not entire arrays).

There is also the matter of different performance characteristics at both ingest and query time for these columns. Ingestion time segment merge is pretty heavy right now because the global value dictionary is stored in heap. Query performance can vary a fair bit with nested columns compared to flat columns, especially with numbers due to the existence of indexes on these numeric columns, which currently at least sometimes results in dramatically faster but also sometimes slower query performance. I'm still exploring this quite a bit, besides documentation follow-up I also have been working on doing some benchmarking to see where things currently stand and plan on sharing those results relatively soon.

So, long story short, due to the unknowns I think the answer for right now is that operators should experiment with `COMPLEX<json>` columns to see if they work well for their use case, and use them in production if so, otherwise provide feedback so that we can continue to make improvements and expand the use cases this is good for?

**Comment by clintropolis** [comment]
I created some query benchmarks to see how various nested column queries perform against traditional columns and see where things currently stand to help guide which areas I will be investigating in the future. I'll add the benchmark code in a follow-up PR. The nested columns are identical to the traditional columns, created using a transform expression:
```
json_object('long1', long1, 'nesteder', json_object('string1', string1, 'long2', long2, 'double3',double3))
```

For the most part things are pretty similar as expected, and the ones that aren't I will be digging deeper into. For numeric columns, which have indexes in nested columns, the indexes in most cases make the queries for nested columns faster, though in a few cases are currently slower. One of the double columns with the bound filter, which part of the reason for that is that currently there is no native numeric range index for the bound filter to use, so bound filters on numbers have to use a `DruidPredicateIndex` which has to match against every value in the dictionary instead of being able to take short-cuts like `LexicographicalRangeIndex` does for strings. I'll be adding this in a follow-up PR as well.

```
SELECT SUM(long1) FROM foo
SELECT SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)) FROM foo
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql        0           5000000        false  avgt    5   36.711 ±  0.917  ms/op
SqlNestedDataBenchmark.querySql        0           5000000        force  avgt    5   15.587 ±  0.276  ms/op
SqlNestedDataBenchmark.querySql        1           5000000        false  avgt    5   39.224 ±  0.870  ms/op
SqlNestedDataBenchmark.querySql        1           5000000        force  avgt    5   15.877 ±  0.440  ms/op


SELECT SUM(long1), SUM(long2) FROM foo
SELECT SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)), SUM(JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT)) FROM foo
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql        2           5000000        false  avgt    5   63.805 ±  1.036  ms/op
SqlNestedDataBenchmark.querySql        2           5000000        force  avgt    5   30.381 ±  1.201  ms/op
SqlNestedDataBenchmark.querySql        3           5000000        false  avgt    5   66.660 ±  0.806  ms/op
SqlNestedDataBenchmark.querySql        3           5000000        force  avgt    5   30.341 ±  1.124  ms/op


SELECT SUM(long1), SUM(long2), SUM(double3) FROM foo
SELECT SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)), SUM(JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT)), SUM(JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE)) FROM foo
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql        4           5000000        false  avgt    5   78.570 ±  1.657  ms/op
SqlNestedDataBenchmark.querySql        4           5000000        force  avgt    5   37.777 ±  1.295  ms/op
SqlNestedDataBenchmark.querySql        5           5000000        false  avgt    5   82.672 ±  1.010  ms/op
SqlNestedDataBenchmark.querySql        5           5000000        force  avgt    5   37.887 ±  0.802  ms/op


SELECT string1, SUM(long1) FROM foo GROUP BY 1 ORDER BY 2,
SELECT JSON_VALUE(nested, '$.nesteder.string1'), SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)) FROM foo GROUP BY 1 ORDER BY 2,
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql        6           5000000        false  avgt    5  269.560 ±  1.454  ms/op
SqlNestedDataBenchmark.querySql        6           5000000        force  avgt    5  157.090 ±  4.058  ms/op
SqlNestedDataBenchmark.querySql        7           5000000        false  avgt    5  373.162 ±  2.871  ms/op
SqlNestedDataBenchmark.querySql        7           5000000        force  avgt    5  195.213 ±  1.993  ms/op


SELECT string1, SUM(long1), SUM(double3) FROM foo GROUP BY 1 ORDER BY 2
SELECT JSON_VALUE(nested, '$.nesteder.string1'), SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)), SUM(JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE)) FROM foo GROUP BY 1 ORDER BY 2
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql        8           5000000        false  avgt    5  251.743 ±  6.438  ms/op
SqlNestedDataBenchmark.querySql        8           5000000        force  avgt    5  172.322 ± 14.814  ms/op
SqlNestedDataBenchmark.querySql        9           5000000        false  avgt    5  417.454 ± 21.276  ms/op
SqlNestedDataBenchmark.querySql        9           5000000        force  avgt    5  215.228 ±  9.304  ms/op


SELECT SUM(long1) FROM foo WHERE string1 = '10000' OR string1 = '1000'
SELECT SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.string1') = '10000' OR JSON_VALUE(nested, '$.nesteder.string1') = '1000'
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       10           5000000        false  avgt    5   11.482 ±  0.495  ms/op
SqlNestedDataBenchmark.querySql       10           5000000        force  avgt    5   11.549 ±  0.303  ms/op
SqlNestedDataBenchmark.querySql       11           5000000        false  avgt    5   11.695 ±  0.293  ms/op
SqlNestedDataBenchmark.querySql       11           5000000        force  avgt    5   11.931 ±  0.338  ms/op


SELECT SUM(long1) FROM foo WHERE long2 = 10000 OR long2 = 1000
SELECT SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) = 10000 OR JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) = 1000
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       12           5000000        false  avgt    5   78.895 ±  2.158  ms/op
SqlNestedDataBenchmark.querySql       12           5000000        force  avgt    5   48.814 ±  0.874  ms/op
SqlNestedDataBenchmark.querySql       13           5000000        false  avgt    5    1.297 ±  0.008  ms/op
SqlNestedDataBenchmark.querySql       13           5000000        force  avgt    5    1.277 ±  0.011  ms/op


SELECT SUM(long1) FROM foo WHERE double3 < 10000.0 AND double3 > 1000.0
SELECT SUM(JSON_VALUE(nested, '$.long1' RETURNING BIGINT)) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE) < 10000.0 AND JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE) > 1000.0
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       14           5000000        false  avgt    5   92.982 ±  1.473  ms/op
SqlNestedDataBenchmark.querySql       14           5000000        force  avgt    5   54.729 ±  0.429  ms/op
SqlNestedDataBenchmark.querySql       15           5000000        false  avgt    5  580.472 ± 28.064  ms/op
SqlNestedDataBenchmark.querySql       15           5000000        force  avgt    5  561.494 ± 54.096  ms/op


SELECT long1, SUM(double3) FROM foo WHERE string1 = '10000' OR string1 = '1000' GROUP BY 1 ORDER BY 2
SELECT JSON_VALUE(nested, '$.long1' RETURNING BIGINT), SUM(JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE)) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.string1') = '10000' OR JSON_VALUE(nested, '$.nesteder.string1') = '1000' GROUP BY 1 ORDER BY 2
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       16           5000000        false  avgt    5  129.760 ±  9.953  ms/op
SqlNestedDataBenchmark.querySql       16           5000000        force  avgt    5  133.015 ± 20.961  ms/op
SqlNestedDataBenchmark.querySql       17           5000000        false  avgt    5  142.197 ±  8.773  ms/op
SqlNestedDataBenchmark.querySql       17           5000000        force  avgt    5  132.048 ± 15.546  ms/op


SELECT string1, SUM(double3) FROM foo WHERE long2 < 10000 AND long2 > 1000 GROUP BY 1 ORDER BY 2
SELECT JSON_VALUE(nested, '$.nesteder.string1'), SUM(JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE)) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) < 10000 AND JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) > 1000 GROUP BY 1 ORDER BY 2
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       18           5000000        false  avgt    5  161.445 ±  7.024  ms/op
SqlNestedDataBenchmark.querySql       18           5000000        force  avgt    5  138.212 ± 19.673  ms/op
SqlNestedDataBenchmark.querySql       19           5000000        false  avgt    5  123.486 ±  5.029  ms/op
SqlNestedDataBenchmark.querySql       19           5000000        force  avgt    5  120.079 ±  6.822  ms/op


SELECT string1, SUM(double3) FROM foo WHERE double3 < 10000.0 AND double3 > 1000.0 GROUP BY 1 ORDER BY 2
SELECT JSON_VALUE(nested, '$.nesteder.string1'), SUM(JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE)) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE) < 10000.0 AND JSON_VALUE(nested, '$.nesteder.double3' RETURNING DOUBLE) > 1000.0 GROUP BY 1 ORDER BY 2
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       20           5000000        false  avgt    5  280.393 ± 15.369  ms/op
SqlNestedDataBenchmark.querySql       20           5000000        force  avgt    5  174.545 ±  2.702  ms/op
SqlNestedDataBenchmark.querySql       21           5000000        false  avgt    5  802.647 ± 32.078  ms/op
SqlNestedDataBenchmark.querySql       21           5000000        force  avgt    5  591.274 ± 16.460  ms/op


SELECT long2 FROM foo WHERE long2 IN (1, 19, 21, 23, 25, 26, 46),
SELECT JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) IN (1, 19, 21, 23, 25, 26, 46),
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       22           5000000        false  avgt    5  273.464 ± 15.731  ms/op
SqlNestedDataBenchmark.querySql       22           5000000        force  avgt    5  272.270 ± 20.511  ms/op
SqlNestedDataBenchmark.querySql       23           5000000        false  avgt    5  174.960 ±  1.923  ms/op
SqlNestedDataBenchmark.querySql       23           5000000        force  avgt    5  177.920 ±  4.095  ms/op


SELECT long2 FROM foo WHERE long2 IN (1, 19, 21, 23, 25, 26, 46) GROUP BY 1",
SELECT JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) FROM foo WHERE JSON_VALUE(nested, '$.nesteder.long2' RETURNING BIGINT) IN (1, 19, 21, 23, 25, 26, 46) GROUP BY 1
Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlNestedDataBenchmark.querySql       24           5000000        false  avgt    5  318.280 ±  7.544  ms/op
SqlNestedDataBenchmark.querySql       24           5000000        force  avgt    5  210.866 ± 14.684  ms/op
SqlNestedDataBenchmark.querySql       25           5000000        false  avgt    5  215.200 ±  2.366  ms/op
SqlNestedDataBenchmark.querySql       25           5000000        force  avgt    5  152.399 ± 22.695  ms/op
```

**Comment by vogievetsky** [comment]
This is amazing stuff. I am excited to follow this up with a web console PR to support some of this stuff there.

**Review by abhishekagarwal87 — COMMENTED** [review]
I had some review comments 😄 Sorry for being late.

File: processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+        {
+          super(name(), args);
+        }
+
+        @Override
+        public ExprEval eval(ObjectBinding bindings)
+        {
+          ExprEval arg = args.get(0).eval(bindings);
+          Object parsed = maybeUnwrapStructuredData(arg);
+          if (arg.type().is(ExprType.STRING) && arg.value() != null && maybeJson(arg.asString())) {
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java:246]: hmm yeah, good point, will adjust in a future PR.

File: processing/src/main/java/org/apache/druid/segment/data/CompressedBlockReader.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+import java.util.function.Supplier;
+
+/**
+ * Reader for a virtual contiguous address range backed by compressed blocks of data.
+ *
+ * Format:
+ * | version (byte) | compression (byte) | num blocks (int) | block size (int) | end offsets | compressed data |
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/segment/data/CompressedBlockReader.java:39]: hah oops, good catch

File: processing/src/main/java/org/apache/druid/segment/data/CompressedBlockReader.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    closer.register(decompressedDataBufferHolder);
+    this.decompressedDataBuffer = decompressedDataBufferHolder.get();
+    this.decompressedDataBuffer.clear();
+    this.byteOrder = byteOrder;
+  }
+
+  /**
+   * Get size in bytes of virtual contiguous buffer
+   */
+  public long getSize()
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/segment/data/CompressedBlockReader.java:137]: no, this is the end position of the last uncompressed value so the uncompressed size. The `endOffsetsBuffer` contains the uncompressed end position of every value that is compressed in the blocks. Using the uncompressed start and end position of a value and knowing the size of blocks we can then locate which block(s) the value is contained in to extract it.

I'll try to add some more comments and javadocs to this area so its a bit clearer what is going on

File: processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+      } else if (adapter instanceof QueryableIndexIndexableAdapter) {
+        dimValues = getSortedIndexesFromQueryableAdapter((QueryableIndexIndexableAdapter) adapter, mergedFields);
+      } else {
+        throw new ISE("Unable to merge columns of unsupported adapter %s", adapter.getClass());
+      }
+
+      boolean allNulls = allNull(dimValues.getSortedStrings()) &&
+                         allNull(dimValues.getSortedLongs()) &&
+                         allNull(dimValues.getSortedDoubles());
+      sortedLookup = dimValues;
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java:106]: nah, this is a strange flow that i picked up from this code being based on `DictionaryEncodedColumnMerger`. Every iteration sets sortedLookup, which is only used if there is a single thing being "merged", which _really_ means that an incremental index is being persisted. If there are more than 1 set things being merged, then it uses a `DictionaryMergingIterator` for each locally sorted set of values to merge them, which null values in the array are just skipped.

Its probably worth adjusting this code, will try to re-arrange in the future so its a bit easier to follow.

File: processing/src/main/java/org/apache/druid/segment/data/CompressedBlockReader.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/segment/data/CompressedBlockReader.java:137]: that will be of help. Since At line #75, it says `final int compressedSize = offsetView.get(numBlocks - 1);`

File: processing/src/main/java/org/apache/druid/segment/data/CompressedBlockSerializer.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    currentOffset += compressed.remaining();
+    offsetValueConverter.clear();
+    offsetValueConverter.putInt(currentOffset);
+    offsetValueConverter.flip();
+    Channels.writeFully(headerOut, offsetValueConverter);
+    Channels.writeFully(valuesOut, compressed);
+    uncompressedDataBuffer.clear();
+    numBlocks++;
+    if (numBlocks < 0) {
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/segment/data/CompressedBlockSerializer.java:164]: how does this happen?
